### PR TITLE
Data team airflow migration

### PR DESCRIPTION
**Pull request by tmccartan**

### CODE REVIEW

File: _posts/2018-03-05-airing-out-a-new-job-system.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+## Next steps
+
+Now that we have our Airflow setup and jobs migrated we are experiencing the full benefits of what airflow has to offer. 
+Within a day we were able to create plugins that allow us to integrate with AWS EMR. This allowed the creation of DAG’s that will in the future support the migration of our ELT process to ETL using Spark, a big 2018 and 2019 initiative here at Gilt/HBC. 
+We are eagerly awaiting the improved DAG deployment system thats coming to Airflow to improve our deployment process but we feel we are in a good place with it at the moment.
+
+The Data teams roadmap has multiple exciting challenges to solve from ingestion, transformation to loading. 
+Having our Airflow setup, we feel we are in a good position now to tackle these problems.
+
+A big callout to [Daniel Matheus Pires](https://github.com/dmateusp) who acted as our go to person in all things Airflow flow.
```
Comment (dmateusp) [inline _posts/2018-03-05-airing-out-a-new-job-system.md]: Mateus without the 'H' ^^ common mistake :)

File: _posts/2018-03-05-airing-out-a-new-job-system.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+- tech
+- data
+- etl
+- scheduling
+---
+
+# Airing out a new Job system
+
+In this article I’ll be sharing some of the knowledge the Data team at Gilt picked up in replacing our old job system with Apache Airflow. 
+We undertook the decision to overhaul our job orchestration system a few months ago due to a number of reason but have now successfully migrated all our data ingestion jobs to the new system.
```
Comment (dmateusp) [inline _posts/2018-03-05-airing-out-a-new-job-system.md]: not sure it's worth mentioning but all our legacy ETL is still on AWS Data pipelines as it was decided to only migrate Dung-beetle and to wait for the Spark switch before migrating the T of the ELT (SQL transformations) there

**Comment by kevinoriordan** [comment]
I think the most interesting part of the story is the details of how you guys extended Airflow to support Batch and if I'm not mistaken, also EMR? Doesn't seem to be mentioned here.

File: _posts/2018-03-05-airing-out-a-new-job-system.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+This lead to some problems with our Sundial job system due to the fact that at the time it didn’t support AWS Batch. 
+This has changed since then with some great work by our personalisation team which you can read more about [here](http://tech.hbc.com/2017-08-04-sundial-batch.html).
+
+## Time for something new…
+
+The team decided it was time to take a fresh approach to how we ran our jobs, which kicked off the investigation of a new system. 
+The investigation took place over a few weeks and spanned across a number of open source solutions. 
+We outlined that a new job system should at minimum support a number of features
+
+* It should be able to integrate with AWS Batch
```
Comment (dmateusp) [inline _posts/2018-03-05-airing-out-a-new-job-system.md]: OR easy to extend and add the feature

**Comment by tmccartan** [comment]
@kevinoriordan  your not mistaken,

File: _posts/2018-03-05-airing-out-a-new-job-system.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+## The refinement…
+
+In the first attempt we figured out lot of the early problems with Airflow but we decided that maintaining the forked version was going to cause trouble down the road.
+To try fix this, we decided to get a vanilla based Airflow instance up and running. This was partly to help with maintaining the system going forward since we didn’t want to have to constantly merge changes from the master branch into our fork.
+This decision coupled with our earlier decision to change some of Airflows base code caused some problems for us.
+
+We decided to go with the latest release of Airflow (1.9) and create a fully dockerised version of Airflow with our DAG’s.
+Luckily there are people already working in this space and we were able to source a lot of the work required from this [repository](https://github.com/puckel/docker-airflow).  
+To migrate the base code that we changed, we found out that Airflow had already solved this problems via it’s [Plugin system](https://airflow.incubator.apache.org/plugins.html).  
+The plugin system gives us a nice way of expanding our functionality and hopefully releasing our plugins as an open sourced in the future.
```
Comment (tmccartan) [inline _posts/2018-03-05-airing-out-a-new-job-system.md]: @kevinoriordan  your not mistaken, we do call out here but I wasn't sure if we wanted to share that. (the general idea is that we would in some day opensource our plugins from here ( https://github.com/gilt/docker-airflow/tree/master/plugins/hbc_plugins)

File: _posts/2018-03-05-airing-out-a-new-job-system.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+# Airing out a new Job system
+
+In this article I’ll be sharing some of the knowledge the Data team at Gilt picked up in replacing our old job system with Apache Airflow. 
+We undertook the decision to overhaul our job orchestration system a few months ago due to a number of reason but have now successfully migrated all our data ingestion jobs to the new system.
+
+Firstly a little bit about our team. The Data Team at Gilt is responsible for constructing, delivering, and supporting all systems and services which enable self-service analytics
+and data science across all banners and all functions at Hudson's Bay Corp (HBC).
+The scope of responsibilities begins with data ingestion, and ends with supporting our user community of BI tool uses and data scientists. 
+These varieties of sources and types of processing lead us to review our job orchestration and I’ll be sharing the challenges involved, attempted solutions and lessons learned.
```
Comment (dmateusp) [inline _posts/2018-03-05-airing-out-a-new-job-system.md]: we reviewed Sundial as an orchestration system for Dung-beetle when we decided to change the back-end to AWS Batch, although Data pipelines (where the T of the ELT resides) also has lots of issues and we considered migrating it to Airflow as well

File: _posts/2018-03-05-airing-out-a-new-job-system.md (hunk 4 above)
Comment (dmateusp) [reply _posts/2018-03-05-airing-out-a-new-job-system.md]: @kevinoriordan when we added AWS Batch the plug in didn't exist (but then it was added so we moved back to the contribution plug-in and improved on it)
Concerning the EMR plug in, we created one as well but, same case, I spotted it on the contributions and we swapped back to it

I think it's worth mentioning that we created an SNS and a Lambda plug-in for example, and AWS Batch / EMR before it was added to the main repo and it was very easy to extend

File: _posts/2018-03-05-airing-out-a-new-job-system.md (hunk 1 above)
Comment (tmccartan) [reply _posts/2018-03-05-airing-out-a-new-job-system.md]: +1

File: _posts/2018-03-05-airing-out-a-new-job-system.md (hunk 2 above)
Comment (tmccartan) [reply _posts/2018-03-05-airing-out-a-new-job-system.md]: Probably oversharing the details, agree this is a bit hand-wavey but it migrating dung-beetle jobs is the majority of the ingestion

File: _posts/2018-03-05-airing-out-a-new-job-system.md (hunk 3 above)
Comment (tmccartan) [reply _posts/2018-03-05-airing-out-a-new-job-system.md]: I think I cover that in the 3rd bullet point

File: _posts/2018-03-05-airing-out-a-new-job-system.md (hunk 4 above)
Comment (tmccartan) [reply _posts/2018-03-05-airing-out-a-new-job-system.md]: I kinda talk about that point in the Next Steps section around the EMR operator but I could elaborate more on the POC where you created the AWS Lamda/SNS and Batch operators (even though we only use Batch for the moment.

**Comment by KarlSimon100** [comment]
With all the previously-incorporated changes, this is looking good!  We could refine it endlessly, so at this point, it LGTM.  Great job, @tmccartan !

**Comment by tmccartan** [comment]
Moved to https://github.com/saksdirect/hbc-tech-blog/pull/20

**Pull request closed without merging**